src/EditorFeatures/Core/Implementation/MetadataAsSource/MetadataAsSourceFileService.cs

```diff
@@ -63,16 +70,18 @@ private static string CreateMutexName(string directoryName)
             return "MetadataAsSource-" + directoryName;
         }
 
-        private string GetRootPathWithGuid_NoLock()
+        private string GetRootPathWithGuid_NoLock(bool allowDecompilation)
         {
-            if (_rootTemporaryPathWithGuid == null)
+            ref string rootTemporaryPathWithGuid = ref (allowDecompilation ? ref _rootDecompiledTemporaryPathWithGuid : ref _rootMetadataTemporaryPathWithGuid);
+            ref Mutex mutex = ref (allowDecompilation ? ref _decompiledMutex : ref _metadataMutex);
```

Comment:
i personally hate this :)  but i guess i will have to grow to live with these sorts of patterns :)

Reply:
Yeah, I admire your desire to play with new things, but if we just went back to a single mutex/folder and generated the paths differently, it's just two lines with no fancy tricks whatsoever.

"If you write your code as cleverly as possible..."